// 100% auto-generated by Source Genesys WebApi Front & Back Creator
// copyright © 2000-2025 Menphis - Sistemas Inteligentes
// This file is part of the Source Genesys project                     
'use client';
import { CRUD_CONSTANTS } from '@/app/tools/crud';
import { FuncaoApi, FuncaoApiError } from '../Apis/ApiFuncao';
import { FilterFuncao } from '../Filters/Funcao';
import { IFuncao } from '../Interfaces/interface.Funcao';
import { FuncaoEmpty } from '../../Models/Funcao';

export class FuncaoValidator {
  static validateFuncao(funcao: IFuncao): { isValid: boolean; errors: string[] } {
    const errors: string[] = [];

    // Atualmente não há validações de regras de negócio específicas
    // Todas as validações são feitas nos inputs correspondentes
    
    return {
      isValid: errors.length === 0,
      errors
    };
  }
}

export interface IFuncaoService {
  fetchFuncaoById: (id: number) => Promise<IFuncao>;
  saveFuncao: (funcao: IFuncao) => Promise<IFuncao>;  
  getList: (filtro?: FilterFuncao) => Promise<IFuncao[]>;
  getAll: (filtro?: FilterFuncao) => Promise<IFuncao[]>;
  deleteFuncao: (id: number) => Promise<void>;
  validateFuncao: (funcao: IFuncao) => { isValid: boolean; errors: string[] };
}

export class FuncaoService implements IFuncaoService {
  constructor(private api: FuncaoApi) {}

  async fetchFuncaoById(id: number): Promise<IFuncao> {
    if (id <= 0) {
      throw new FuncaoApiError('ID inválido', 400, 'INVALID_ID');
    }

    try {
      
      const response = await this.api.getById(id);
      return response.data;

    } catch (error) {
      if (error instanceof FuncaoApiError) {
        throw error;
      }
      throw new FuncaoApiError('Erro ao buscar funcao', 500, 'FETCH_ERROR', error);
    }
  }

  async saveFuncao(funcao: IFuncao): Promise<IFuncao> {    
    const validation = this.validateFuncao(funcao);
    if (!validation.isValid) {
      throw new FuncaoApiError(
        `Dados inválidos: ${validation.errors.join(', ')}`,
        400,
        'VALIDATION_ERROR'
      );
    }

    try {
      const response = await this.api.addAndUpdate(funcao);
      return response.data;
    } catch (error) {
      if (error instanceof FuncaoApiError) {
        throw error;
      }
      throw new FuncaoApiError('Erro ao salvar funcao', 500, 'SAVE_ERROR', error);
    }
  }

  
    async getList(filtro?: FilterFuncao): Promise<IFuncao[]> {
    try {
      const response = await this.api.getListN(CRUD_CONSTANTS.MAX_RECORDS_COMBO, filtro);
      return response.data || [];
    } catch (error) {
        if (process.env.NEXT_PUBLIC_SHOW_LOG === '1')
        console.log('Error fetching Funcao list');
      return [];
    }
  }

 
  

   async getAll(
    filtro?: FilterFuncao,
    onOnlineData?: (data: IFuncao[]) => void
  ): Promise<IFuncao[]> {
    try {
      // Carrega dados offline primeiro
      const preloadResponse = await this.api.filterPreload(0, filtro ?? {});
      const offlineData = preloadResponse?.data || [];

      if (onOnlineData) {
        // Busca dados online em background e envia via callback
        this.api.filter(0, filtro ?? {})
          .then(response => {
            if (response?.data) {
              onOnlineData(response.data);
            }
          })
          .catch(error => {
            if (process.env.NEXT_PUBLIC_SHOW_LOG === '1')
                console.log('Error fetching online Funcao');
          });
        // Retorna offline imediatamente
        return offlineData;
      } else {
        // Se não há callback, aguarda dados online e retorna
        try {
          const onlineResponse = await this.api.filter(0, filtro ?? {});
          return onlineResponse?.data || offlineData;
        } catch (error) {
            if (process.env.NEXT_PUBLIC_SHOW_LOG === '1')
                console.log('Error fetching online Funcao');
          return offlineData;
        }
      }
    } catch (error) {
      if (process.env.NEXT_PUBLIC_SHOW_LOG === '1')
        console.log('Error fetching all Funcao:');
      return [];
    }
  }


  async deleteFuncao(id: number): Promise<void> {
    if (id <= 0) {
      throw new FuncaoApiError('ID inválido para exclusão', 400, 'INVALID_ID');
    }

    try {
      await this.api.delete(id);
    } catch (error) {
      if (error instanceof FuncaoApiError) {
        throw error;
      }
      throw new FuncaoApiError('Erro ao excluir funcao', 500, 'DELETE_ERROR', error);
    }
  }

  validateFuncao(funcao: IFuncao): { isValid: boolean; errors: string[] } {
    return FuncaoValidator.validateFuncao(funcao);
  }
}